documentation/public/internals/design-docs/language-definition-v2/index.md

```diff
@@ -0,0 +1,438 @@
+
+For the CST, I think the current algorithms work well, and we should be able to keep them. Unrecognized (skipped) input
+is grouped into one token, and we can just add it as-is to the `cst` node under its AST node.
+
+For the AST, I think we can generate/implement a `GenerateMissing` trait for each type. If parsing fails at any point, and
+there are remaining fields in the parser that need constructing, we can call it as we are unwinding the stack. It will:
+
+-   For terminals, generate a token containing an empty string value and `TerminalKind::MISSING`.
+-   For repeated/separated items, create an empty `Vec<>`.
+-   For enums, it will create an instance of the variant specified in `default_variant`.
+-   For expressions, it will create an instance of the expression specified in `default_primary_expression`.
+-   For other non-terminal fields, call the same trait recursively.
+
+This also plays well with existing CST error-recovery mechanisms we have in place. By having the untyped cst node
+in each AST type, it can be incomplete if there are errors, and have `TerminalKind::UNRECOGNIZED` nodes, which is completely
+fine. Users can get and iterate on that, and also reconstruct the complete input from the raw input tokens. The
+`TerminalKind::MISSING` tokens don't need to be added to the CST at all, so that it remains true to input.
+
+## Public API Changes
+
+Based on the above, I propose the following changes to the current public API:
+
+-   Rename `TokenKind` to `TerminalKind`, since it will also refer to trivia.
+-   Rename `RuleKind` to `NonTerminalKind`, since "rule" is ambiguous.
+-   Rename `TerminalKind::SKIPPED`to `UNRECOGNIZED` for clarity, and add `TerminalKind::MISSING` for partial AST nodes.
+-   Hide `LexicalContext` and `fn scan(TokenKind)` from the public API, as it is a short-term workaround,
+    and will be replaced later when we have language embedding.
+-   Replace `fn parse(NonTerminalKind)` with a set of typed functions `fn parse_foo() -> ParseResult<Foo>` that return
+    AST nodes directly. Users can still call `foo_node.cst` to get the CST if needed.
+-   Remove `ProductionKind` completely, since it is no longer needed.
+-   Since `EndOFFileTrivia` no longer exists, `fn parse_foo() -> ParseResult<Foo>` should collect any remaining trivia
+    at the end of the input, and include it in the `ParseResult` returned, for any kind of node, not just `SourceUnit`.
+
+## Visitors and Cursors
+
+The current CST visitors/cursors should still work as-is, since the CST tree will be unchanged. However, the new AST types
+allow us in the future to produce typed visitors and traits with named functions for every node type, similar to a lot of
+other AST processing libraries. I want to at least produce an immutable `Visitor` and a mutable `Rewriter`.
+
+## Open Questions
+
+-   How would terminators and delimiters (the new `error_recovery` field) affect this? Is it possible to still use it
```

Comment:
As above, I think we should totally ignore this for now. Come back to it in a few quarters.

Reply:
Removed.